### PR TITLE
Weighted Mixture of Datasets for Training

### PR DESCRIPTION
**Pull request by Ivan-Zhou**
This PR implements https://github.com/stanford-crfm/levanter/issues/218. 
- [x] Design doc
- [x] Implement LMMixtureDatasetConfig 
- [x] Implement MixtureDataset 
- [x] Testing

### CODE REVIEW

File: docs/design/Multiple-Data-Mixture.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+a for loop over `one_item_generator`, then at every step, we slice from `one_item_generator` to get 
+B sequence, what do we want to achieve with the first for loop?
+
+### Weighted Sampling
+At every step of `ShardedBatchLoader.__iter__()`, we need to sample B sequences from N datasets. 
+The weights of the datasets are used to sample from N datasets. Specifically, we will sample
+B sequences from the i-th dataset with probability `weight[i] / sum(weight)`. 
+
+### Deterministic Batches
+I am not sure if we can still achieve deterministic batches with weighted sampling.
```
Comment (dlwh) [inline docs/design/Multiple-Data-Mixture.md:86]: as long as the random numbers are deterministic it's fine.

supporting this with #119 is hard though

File: docs/design/Multiple-Data-Mixture.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+### Validation Set
+We will apply the same logic to `ReplicatedBatchLoader` for validation sets.
```
Comment (dlwh) [inline docs/design/Multiple-Data-Mixture.md:89]: i think it's better to keep validation sets split and report separately

File: docs/design/Multiple-Data-Mixture.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+B sequences from the i-th dataset with probability `weight[i] / sum(weight)`. 
+
+### Deterministic Batches
+I am not sure if we can still achieve deterministic batches with weighted sampling.
+
+### Validation Set
+We will apply the same logic to `ReplicatedBatchLoader` for validation sets. 
+
+*TODO: remove the following question*
+> Question: right now does Levanter supports a dataset that only has a training set but no validation set?
```
Comment (dlwh) [inline docs/design/Multiple-Data-Mixture.md:92]: not explicitly, though I think it might just work if you use 0 urls

File: docs/design/Multiple-Data-Mixture.md [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+1. Users can specify the weight of each domain in the dataset, through the configuration file.
+2. At every batch, the weight of each domain is used to sample the number of tokens from the 
+corresponding domain.
+
+Ideally, we still want to preserve the reproducibility and deterministic batches of Levaner, 
+as specified in the [Data Loader design](Data-Loader-Design.md).
+
+
+## Design and Implementation
+### Configuration
```
Comment (dlwh) [inline docs/design/Multiple-Data-Mixture.md:30]: this section looks great

File: docs/design/Multiple-Data-Mixture.md [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+      cache_dir: "gs://pubmed-mosaic/tokenized/reddit/"
+```
+
+Note that this design is backward compatible, as users can still specify a single dataset in the
+configuration file, and the weight of the dataset will be 1.0 by default.
+
+### Mixing at Batch Level
+Currently at training, `ShardedBatchLoader` takes in a single dataset of TokenSeqDataset class, 
+and implements the batching and sharding logic. Therefore, the mixing of multiple datasets should 
+be implemented at ShardedBatchLoader.
```
Comment (dlwh) [inline docs/design/Multiple-Data-Mixture.md:70]: I'd rather we do this at the dataset level: have a MixtureDataset that takes a list of datasets and weights and does the logic you describe. feels like a better separation of concerns

File: docs/design/Multiple-Data-Mixture.md [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+- The [Pile dataset](https://pile.eleuther.ai/) is a mixture of 22 different datasets, including
+web data (24%), Wikipedia (9%), GitHub (4%), and more;
+- The [RedPajama dataset](https://huggingface.co/datasets/togethercomputer/RedPajama-Data-1T) is 
+a mixture of 1.2 tokens, from domains such as CommonCrawl (73%), GitHub (5%), Books (2.4%), 
+Wikipedia (2.4%), etc. 
+- The [Falcon 40B](https://huggingface.co/tiiuae/falcon-40b) model is trained on 1T tokens from 
+sources including RedfinedWeb English (75%), Books (6%), Conversations (5%), etc. 
+
+The common approach at training, used in LLaMA, MPT, and Falcon models, is to sample from each domain in 
+the dataset uniformly, based on the percentage of tokens in each domain. Recently, we have seen a new line
```
[the remark this answers is not in the document]
Comment (Ivan-Zhou) [reply docs/design/Multiple-Data-Mixture.md:16]: You are right: 
image

File: src/levanter/data/text.py [hunk 7]
```diff
@@ -107,6 +126,60 @@ def load(pos: Axis, cache_dir: str, stride: Optional[int] = None) -> "TokenSeqDa
...
+        docs from doc_cache and yield batches of token sequences. We leverage this
+        implementation of iteration. This function mainly samples a dataset according to
+        the weights and call __iter__() function from corresponding TokenSeqDataset.
+
+        Yields:
+            Iterator[NamedArray]: _description_
+        """
+        while True:
+            dataset_index = self.sample_index(self.weights)
+            yield from self.token_seq_datasets[dataset_index]
```
Comment (Ivan-Zhou) [inline src/levanter/data/text.py:167]: Instead of yield from, we should call from a list of iterators

File: src/levanter/data/text.py [hunk 8]
```diff
@@ -107,6 +126,60 @@ def load(pos: Axis, cache_dir: str, stride: Optional[int] = None) -> "TokenSeqDa
...
+    their associated weights.
+    We leverage the implementation of TokenSeqDataset on sharding and iterating over the data.
+
+    :param doc_cache_list: a list of TokenizedDocumentCache to draw from
+    :param Pos: the axis to use for the sequences. Sequences will be a NamedArray with axis Pos
+    :param weight_list: a list of weights for each dataset
+    """
+
+    def __init__(self, doc_caches: List, Pos: Axis, weights: List[float]):
+        self.token_seq_datasets = [TokenSeqDataset(doc_cache, Pos) for doc_cache in doc_caches]
```
Comment (Ivan-Zhou) [inline src/levanter/data/text.py:141]: We can make this more abstract

File: src/levanter/data/text.py [hunk 9]
```diff
@@ -502,8 +562,14 @@ class LMDatasetConfig:
     def the_tokenizer(self) -> PreTrainedTokenizerFast:
         return load_tokenizer(self.tokenizer)
 
-    def build_or_load_cache(self, split: str, monitors: Union[bool, List[MetricsMonitor]] = True):
-        source = self.get_shard_source(split)
+    def build_or_load_cache(
+        self, split: str, monitors: Union[bool, List[MetricsMonitor]] = True
+    ) -> Optional[TokenizedDocumentCache]:
```
Comment (Ivan-Zhou) [inline src/levanter/data/text.py:567]: This is for handling cases when the dev split is not available (RP dataset, for example)

**Comment by Ivan-Zhou** [comment]
@dlwh While I work on fixing the validation loss = 0, maybe you can start reviewing and let me know if there's any comment

File: src/levanter/data/text.py [hunk 10]
```diff
@@ -175,6 +174,75 @@ def load(seq_len: int, cache_dir: str, stride: Optional[int] = None) -> "TokenSe
...
+        return MixtureDataset(sharded_doc_caches, self.seq_len, self.weights)
+
+    def __iter__(self) -> Iterator[np.ndarray]:
+        """TokenSeqDataset has a non-trivial implementation of __iter__() that iterates
+        docs from doc_cache and yield batches of token sequences. We leverage this
+        implementation of iteration. This function mainly samples a dataset according to
+        the weights and call __iter__() function from corresponding TokenSeqDataset.
+
+        Edge cases to consider:
+        1. Some of the iterators have been exhausted. We need to drop them and corresponding weights.
```
[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/data/text.py:221]: we should make take inspiration from https://huggingface.co/docs/datasets/v2.14.0/en/package_reference/main_classes#datasets.interleave_datasets

File: src/levanter/data/text.py [hunk 11]
```diff
@@ -580,6 +657,44 @@ def get_shard_source(self, split) -> ShardedDataSource[str]:
...
+    datasets: List[LMDatasetConfig]
+
+    def __post_init__(self):
+        if len(self.datasets) == 0:
+            raise ValueError("Must provide at least one dataset")
+
+        self.weights = [d.weight for d in self.datasets]
+
+        # ensure all datasets use the same tokenizer
+        if len(set([d.tokenizer for d in self.datasets])) != 1:
```
Comment (dlwh) [inline src/levanter/data/text.py:675]: maybe we shouldn't use LMDatasetConfig then.

What about LMDatasetSourceConfig that is just (splits of) URLs or a hf dataset it, then LMDatasetConfig takes a list of those (or map) along with weights as well as a tokenizer?

I'm not super concerned about making sure the configs are backwards compatible. Our user base is low enough ATM that we can break without making too many people mad, so I'd rather get the right thing while we still can

File: config/gpt2_data_mix.yaml [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    - train_urls:
+          - "gs://pubmed-mosaic/openwebtext-sharded/openwebtext_train.{1..128}-of-128.jsonl.gz"
+      validation_urls:
+        - "gs://pubmed-mosaic/openwebtext-sharded/openwebtext_val.{1..8}-of-8.jsonl.gz"
+      cache_dir: "gs://levanter-data/tokenized/openwebtext/"
+      tokenizer: "gpt2"
+      weight: 0.5
+    - id: dlwh/wikitext_103_detokenized
+      cache_dir: "gs://levanter-data/tokenized/wikitext"
+      weight: 0.5
```
[the remark this answers is not in the document]
Comment (Ivan-Zhou) [reply config/gpt2_data_mix.yaml:12]: Yes I do agree that such interface would be nicer for tuning dataset weights. Is there any particular reason that you assign names for each task? 

Maybe an equivalent config for data mixture looks like this:

```
data-mixture:
  data_configs:
    pile:
      train_urls:
        - gs://levanter-data/pile/train/{00..29}.jsonl.zst
      validation_urls:
        - gs://levanter-data/pile/val.jsonl.zst
      cache_dir: "gs://levanter-data/tokenized/pile-old/"
      tokenizer: "EleutherAI/gpt-neox-20b"
    redpajama:
      id: togethercomputer/RedPajama-Data-1T
      cache_dir: gs://levanter-data/tokenized/redpajama/
      tokenizer: EleutherAI/gpt-neox-20b
      splits:
        - train
      rows_per_chunk: 4096
  data_weights:
    pile: 0.6
    redpajama: 0.4
```

I think this could be more flexible and cleaner than the current approach, which is trying to modify the existing `data` field while ensure compatibility (referring back to your comment about LMDatasetConfig).

File: src/levanter/data/text.py [hunk 13]
```diff
@@ -175,6 +174,75 @@ def load(seq_len: int, cache_dir: str, stride: Optional[int] = None) -> "TokenSe
...
+                yield item
+            except StopIteration:
+                # if the iterator is exhausted, we need to drop the iterator and its weight
+                del token_seq_iterators[dataset_index]
+                del weights[dataset_index]
+
+    @staticmethod
+    def sample_index(weights):
+        """Sample a dataset according to the weights"""
+        return np.random.choice(len(weights), p=weights)
```
[the remark this answers is not in the document]
Comment (Ivan-Zhou) [reply src/levanter/data/text.py:239]: Acknowledge.

File: src/levanter/data/text.py [hunk 14]
```diff
@@ -175,6 +174,75 @@ def load(seq_len: int, cache_dir: str, stride: Optional[int] = None) -> "TokenSe
...
+    """
+
+    def __init__(self, doc_caches: List, seq_len: int, weights: List[float]):
+        doc_caches, weights = self._check_doc_cache(doc_caches, weights)
+        self.doc_caches = doc_caches
+        self.token_seq_datasets = [TokenSeqDataset(doc_cache, seq_len) for doc_cache in doc_caches]
+        self.seq_len = seq_len
+        self.weights = self.normalize_weights(weights)
+
+    def _check_doc_cache(self, doc_caches, weights):
```
[the remark this answers is not in the document]
Comment (Ivan-Zhou) [reply src/levanter/data/text.py:195]: Acknowledge.

**Comment by Ivan-Zhou** [comment]
It is time for me to re-pick up this PR. Given the amount of changes that have been made to the dataset and text, I might need to proceed forward with this PR or make a new one from a clean start. 

I summarized TODOs based on David's comments :
- Make `LMDatasetSourceConfig` or `LMDatasetMixtureConfig` that takes in a list of dataset configs as well as weights. It should support more than one tokenizer and provide a clean interface for tuning dataset weights. 
- Borrow `oversampling` argument and logic from datasets.interleave_datasets. I might need to take a closer look at their implementation. 
- Move `_check_doc_cache` outside of MixtureDataset and expect caches as input to MixtureDataset
- Create an instance rng as a member variable and a seed argument as well.

**Review by dlwh — APPROVED** [review]
